source/common/router/config_impl.h

```diff
@@ -656,22 +662,26 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
     return vhost_.virtualClusterFromEntries(headers);
   }
   std::chrono::milliseconds timeout() const override { return timeout_; }
-  absl::optional<std::chrono::milliseconds> idleTimeout() const override { return idle_timeout_; }
+  absl::optional<std::chrono::milliseconds> idleTimeout() const override {
+    return idle_timeout_.count() >= 0 ? absl::optional(idle_timeout_) : absl::nullopt;
+  }
   bool usingNewTimeouts() const override { return using_new_timeouts_; }
   absl::optional<std::chrono::milliseconds> maxStreamDuration() const override {
-    return max_stream_duration_;
+    return max_stream_duration_.count() >= 0 ? absl::optional(max_stream_duration_) : absl::nullopt;
   }
   absl::optional<std::chrono::milliseconds> grpcTimeoutHeaderMax() const override {
-    return grpc_timeout_header_max_;
+    return grpc_timeout_header_max_.count() >= 0 ? absl::optional(grpc_timeout_header_max_)
+                                                 : absl::nullopt;
   }
   absl::optional<std::chrono::milliseconds> grpcTimeoutHeaderOffset() const override {
-    return grpc_timeout_header_offset_;
+    return grpc_timeout_header_offset_.count() >= 0 ? absl::optional(grpc_timeout_header_offset_)
+                                                    : absl::nullopt;
   }
   absl::optional<std::chrono::milliseconds> maxGrpcTimeout() const override {
-    return max_grpc_timeout_;
+    return max_grpc_timeout_.count() >= 0 ? absl::optional(max_grpc_timeout_) : absl::nullopt;
   }
   absl::optional<std::chrono::milliseconds> grpcTimeoutOffset() const override {
-    return grpc_timeout_offset_;
+    return grpc_timeout_offset_.count() >= 0 ? absl::optional(grpc_timeout_offset_) : absl::nullopt;
```

Comment:
how hard would it be to change call-sites to compare against the Unset constant rather than having to convert to a std::optional and then have another conditional at the call-site?

Reply:
`optional` in the return value of the API makes sense to me -- it clear communicates to the caller that this may or may not have a value. If we just return a duration with a magic sentinel, we lose that clarity. We can convert the current callsites, but any future developer would not expect to have to compare the result to `Unset` constant.

Reply:
fair enough -- that is a clear interface and I'm fine with it.

I was prematurely optimizing that the expanded code would have two conditionals rather than one. What you have is fine.